contracts/src/v0.8/dev/ocr2dr/OCR2DRClient.sol

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+
+import "./OCR2DR.sol";
+import "../interfaces/OCR2DRClientInterface.sol";
+import "../interfaces/OCR2DROracleInterface.sol";
+
+/**
+ * @title The OCR2DR client contract
+ * @dev This contract is similar to ChainlinkClient in many respects.
+ * @notice Contract writers can inherit this contract in order to create on-demand OCR requests
+ */
+abstract contract OCR2DRClient is OCR2DRClientInterface {
+  OCR2DROracleInterface private s_oracle;
+  uint256 private s_requestCount = 1;
+  mapping(bytes32 => address) private s_pendingRequests;
+
+  event RequestSent(bytes32 indexed id);
+  event RequestFulfilled(bytes32 indexed id);
+  event RequestCancelled(bytes32 indexed id);
+
+  error TransferAndCallError();
+  error SenderIsNotOracle();
+  error RequestIsAlreadyPending();
+  error RequestIsNotPending();
+
+  /**
+   * @notice Sends OCR2DR request to the stored oracle address
+   * @param req The initialized OCR2DR.Request
+   * @param subscriptionId The subscription ID
+   * @return requestId The generated request ID
+   */
+  function sendRequest(OCR2DR.Request memory req, uint256 subscriptionId) internal returns (bytes32) {
+    uint256 nonce = s_requestCount;
+    s_requestCount = nonce + 1;
+    bytes32 requestId = keccak256(abi.encodePacked(this, nonce));
```

Comment:
Shouldn't the Oracle contract assign request IDs? What if a malicious client contract re-uses on existing ID to mess up the response?
I also wonder if it could be something computationally-cheaper.

Reply:
This logic is borrowed from the existing Direct Request. I guess the oracle would need to mix client's nonce and the client's address to make up a request ID for DON. I will consider moving this responsibility to Oracle while working on Oracle.